resourcemanager/pool/spool/spool.go

```diff
@@ -0,0 +1,171 @@
+	mu                 sync.Mutex
+	options            *Options
+	capacity           atomic.Int32
+	running            atomic.Int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	pool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	if size == 0 {
+		return nil, pool.ErrPoolParamsInvalid
+	}
+	result.SetName(name)
+	result.capacity.Store(size)
+	result.concurrencyMetrics.Set(float64(size))
+	err := resourcemanager.InstanceResourceManager.Register(result, name, component)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Tune changes the pool size.
+func (p *Pool) Tune(size int) {
+	if size == 0 {
+		return
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.SetLastTuneTs(time.Now())
+	p.capacity.Store(int32(size))
+	p.concurrencyMetrics.Set(float64(size))
+}
+
+// Run runs a function in the pool.
+func (p *Pool) Run(fn func()) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	_, run := p.checkAndAddRunning(1)
+	if !run {
+		return pool.ErrPoolOverload
+	}
+	p.taskManager.RegisterTask(p.GenTaskID(), 1)
+	p.run(fn)
+	return nil
+}
+
+// Cap returns the capacity of the pool.
+func (p *Pool) Cap() int {
+	return int(p.capacity.Load())
+}
+
+// Running returns the number of running goroutines.
+func (p *Pool) Running() int {
+	return int(p.running.Load())
+}
+
+func (p *Pool) run(fn func()) {
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("recover panic", zap.Any("recover", r))
+			}
+			p.wg.Done()
+			p.running.Add(-1)
+		}()
+		fn()
+	}()
+}
+
+// RunWithConcurrency runs a function in the pool with concurrency.
+func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	conc, run := p.checkAndAddRunning(concurrency)
+	if !run {
+		return pool.ErrPoolOverload
+	}
+	// TODO: taskManager need to refactor
+	p.taskManager.RegisterTask(p.GenTaskID(), conc)
+	for n := int32(0); n < conc; n++ {
+		p.run(func() {
+			for fn := range fns {
+				fn()
+			}
+		})
+	}
+	return nil
+}
+
+// checkAndAddRunning is to check if a task can run. If can, add the running number.
+func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	for {
+		value, run := p.checkAndAddRunningInternal(int32(concurrency))
+		if run {
+			return value, run
+		}
+		if !p.options.Blocking {
+			return 0, false
+		}
+		time.Sleep(5 * time.Millisecond)
```

Comment:
Can we use a condition variable here? Wake up with `5ms` latency is not desired.

Reply:
It is a good idea. It will be implemented in the future.